### DISCUSSION TITLE
How to impose NeumannBC on a boundary line of a 2D plane in a 3D mesh?

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#16850] [Q&A Meshing] [answered]
Hi everyone,

I use Gmsh (attached .geo and .msh files ) to build a 3D model (named 'Matrix') embedded by a 2D plane (named 'Fault'). I want to impose NeumannBC on the left boundary line (named 'Leftline')of the 2D plane.

I have done the following works:
1) Defined the Fault as a block by using 'LowerDBlockFromSidesetGenerator' in the input file of MOOSE or by inserting 'lower_dimensional_block' after the "Fault" name in the .msh file.

2) Imposed NeumannBC on the  'Leftline' in the input file of MOOSE.

3) Started my simulation, then I got the error information:
_The object ''Leftline_h' contains the following side set ids that do not exist on the mesh: 1. MOOSE distinguishes between "node sets" and "side sets" depending on whether you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding to your side sets are constructed for you by default. Try setting "Mesh/construct_side_set_from_node_set=true" if you see this error._

I setted _construct_side_set_from_node_set=true_  but I got the same error information again.  But if I set DirichletBC on the 'Leftline', this was no error. 

Could you please give me suggestions on this? Thanks a lot.

Yan
![case](https://user-images.githubusercontent.com/56687392/106612182-1ef0c880-6569-11eb-8e9d-3dc2d24ad77f.png)

[mesh files.zip](https://github.com/idaholab/moose/files/5911703/mesh.files.zip)

**Comment by hugary1995** [comment, marked as answer]
Could you also attach your input file (at least the [Mesh] block) after adding "construct_side_set_from_node_set=true"?

**Reply by Joseph-0123** [reply]
Yes, here it is.

[Mesh]
  [file]
    type = FileMeshGenerator
    file = case.msh
  []
  [./Faultld]
    input = file
    type = LowerDBlockFromSidesetGenerator
    new_block_id = '10'
    new_block_name = 'Faultld'
    sidesets = '2'
  [../]
  construct_side_list_from_node_list=true
[]

[BCs]
  [./Leftline_h]
    type = NeumannBC
    variable = pressure
    boundary = 'Leftline'
    value = -1e-5
 [../]
[]

**Reply by hugary1995** [reply]
Should it be "construct_side_list_from_node_list" or "construct_side_set_from_node_set"?

**Reply by Joseph-0123** [reply]
Actually, I tried both two options (using together or using just one) and all failed.

**Reply by hugary1995** [reply]
When I run it with "--mesh-only", and it doesn't show any side set or node set in it. Indeed, there are some empty nodes in "case.msh". If I were you I would double check my gmsh journal.

**Reply by Joseph-0123** [reply]
Gary, thanks for your remind.  The Gmsh journal is attached here. 

In my case,  defining the 2D Fault as a block From Sidesets to Nodesets is valid. But defining 1D 'Leftline'  as a boundary using the  "From Nodesets to Sidesets" has an error and maybe conflict with "From Sidesets to Nodesets". Because in one MOOSE simulation, only one way (“From Sidesets to Nodesets” or"From Nodesets to Sidesets") can be used and can not be used together(see https://mooseframework.inl.gov/source/mesh/MooseMesh.html). 

I don't know whether my former understanding is right or wrong. Please correct me, thanks.


[journal.txt](https://github.com/idaholab/moose/files/5914673/journal.txt)

**Reply by GiudGiud** [reply]
Has this been resolved?

**Reply by saturn00000** [reply]
> Has this been resolved?

Hi @GiudGiud,

can we apply NeumannBC on **a single node** in 2D?

I tried to test NeumannBC on **a single node** (named 'ptt') of the 2D plane (the mesh is generated by Gmsh).
In order to use this gmsh file, **construct_side_list_from_node_list = true** is **active** in MOOSE(see https://github.com/idaholab/moose/discussions/16850#discussioncomment-331888 ).

But **construct_side_list_from_node_list = true** cannot be used due to the error:
```
(BCs/right/boundary):
    the following side sets (ids) do not exist on the mesh: ptt (10)

    MOOSE distinguishes between "node sets" and "side sets" depending on whether
    you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
    to your side sets are constructed for you by default.

    Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
    Note: If you are running with adaptivity you should prefer using side sets.
```

Here is the input file
```
[Mesh]
  [file]
  type = FileMeshGenerator
  file = gmsh/model04.msh
  []

construct_node_list_from_side_list = false # true
construct_side_list_from_node_list = true
[]


[Variables]
  [./u]
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]
[]

[BCs]
  [./right]
    type = NeumannBC
    variable = u
    boundary = ptt
    value = 2
  [../]
[]

[Executioner]
  type = Steady

  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
[]
```

.msh file:

```
$MeshFormat
4.1 0 8
$EndMeshFormat
$PhysicalNames
1
0 10 "ptt"
$EndPhysicalNames
$Entities
4 4 1 0
1 0 0 0 0 
2 10 0 0 0 
3 10 10 0 1 10 
4 0 10 0 0 
1 -1.000000002804313e-07 -1e-07 -1e-07 10.0000001 1e-07 1e-07 0 2 1 -2 
2 9.999999900000001 -1.000000002804313e-07 -1e-07 10.0000001 10.0000001 1e-07 0 2 2 -3 
3 -1.000000002804313e-07 9.999999900000001 -1e-07 10.0000001 10.0000001 1e-07 0 2 3 -4 
4 -1e-07 -1.000000002804313e-07 -1e-07 1e-07 10.0000001 1e-07 0 2 4 -1 
1 -1.000000002804313e-07 -1.000000002804313e-07 -1e-07 10.0000001 10.0000001 1e-07 0 4 4 1 2 3 
$EndEntities
$Nodes
9 9 1 9
0 1 0 1
2
0 0 0
0 2 0 1
3
10 0 0
0 3 0 1
1
10 10 0
0 4 0 1
4
0 10 0
1 1 0 1
5
5 0 0
1 2 0 1
6
10 5 0
1 3 0 1
7
5 10 0
1 4 0 1
8
0 5 0
2 1 0 1
9
5.01650975510978 5.01650975510978 0
$EndNodes
$Elements
9 16 1 16
0 1 15 1
2 2 
0 2 15 1
3 3 
0 3 15 1
1 1 
0 4 15 1
4 4 
1 1 1 2
5 2 5 
6 5 3 
1 2 1 2
7 3 6 
8 6 1 
1 3 1 2
9 1 7 
10 7 4 
1 4 1 2
11 4 8 
12 8 2 
2 1 3 4
13 9 8 2 5 
14 9 5 3 6 
15 9 7 4 8 
16 6 1 7 9 
$EndElements
```

**Reply by GiudGiud** [reply]
Hello

With a nodal variable yes you should be able to.
However, you cannot create a side from a single node. So anything that expects a side to build a boundary condition (like in finite volume for example) will not be able to.

Guillaume
